homeassistant/components/huawei_lte/__init__.py

```diff
@@ -59,98 +118,231 @@
 
+        get_data(KEY_DIALUP_MOBILE_DATASWITCH, self.client.dial_up.mobile_dataswitch)
         get_data(KEY_DEVICE_INFORMATION, self.client.device.information)
+        if self[KEY_DEVICE_INFORMATION]:
+            # Full information includes everything in basic
+            self.unsubscribe(KEY_DEVICE_BASIC_INFORMATION)
+        get_data(KEY_DEVICE_BASIC_INFORMATION, self.client.device.basic_information)
         get_data(KEY_DEVICE_SIGNAL, self.client.device.signal)
         get_data(
             KEY_MONITORING_TRAFFIC_STATISTICS, self.client.monitoring.traffic_statistics
         )
         get_data(KEY_WLAN_HOST_LIST, self.client.wlan.host_list)
 
+    def cleanup(self, *_) -> None:
+        """Clean up resources."""
+        if hasattr(self.client, "user"):
+            try:
+                self.client.user.logout()
+            except ResponseErrorNotSupportedException:
+                _LOGGER.debug("Logout not supported by device", exc_info=True)
+            except ResponseErrorLoginRequiredException:
+                _LOGGER.debug("Logout not supported when not logged in", exc_info=True)
+            except BaseException:
+                _LOGGER.warning("Logout error", exc_info=True)
+
 
 @attr.s
 class HuaweiLteData:
     """Shared state."""
 
-    data = attr.ib(init=False, factory=dict)
+    hass_config = attr.ib()
+    routers = attr.ib(init=False, factory=dict)
+
+    def get_router_data(self, url):
+        """Get the requested router data."""
+        return self.routers.get(url)
+
+
+async def async_setup_entry(hass, config_entry: ConfigEntry):
+    """Set up Huawei LTE component from config entry."""
+
+    router_config = _router_config_from_config_entry(config_entry)
+
+    # See if we already have an entry with this URL
+    for url in (
+        url_normalize(x, default_scheme="http") for x in hass.data[DOMAIN].routers
+    ):
+        if url == router_config[CONF_URL]:  # entry url already normalized
+            _LOGGER.warning(
+                "Router at %s is already configured, ignoring config entry", url
+            )
+            return False
+
+    await hass.async_add_executor_job(_setup_lte, hass, router_config)
+
+    router_data = hass.data[DOMAIN].routers[router_config[CONF_URL]]
+    config_entry.add_update_listener(router_data.async_options_updated)
+
+    device_data = {}
+    sw_version = None
+    if router_data[KEY_DEVICE_INFORMATION]:
+        serial_number = router_data[KEY_DEVICE_INFORMATION].get("SerialNumber")
+        if serial_number:
+            device_data["identifiers"] = {(DOMAIN, serial_number)}
+        sw_version = router_data[KEY_DEVICE_INFORMATION].get("SoftwareVersion")
+    elif router_data[KEY_DEVICE_BASIC_INFORMATION]:
+        sw_version = router_data[KEY_DEVICE_BASIC_INFORMATION].get("SoftwareVersion")
+    if sw_version:
+        device_data["sw_version"] = sw_version
+
+    device_registry = await dr.async_get_registry(hass)
```

Comment:
Please limit this PR to just the required changes to implement config entries. We don't need to implement the device registry.

Reply:
I believe doing this will make it possible to assign the hub and devices connected to it to an area. Why wouldn't that be desirable?

Reply:
It's desirable, but we try to limit the scope of PRs. This is already a big PR. Everything we can move to later PRs is a good move.

Reply:
Done.